Stabilize PA on throughput only for async case

dyastremsky opened the pull request:
After fixing the async request rate case, perf analyzer no longer stabilizes most of the time for smaller models due to the latency being more variable in the async case. As a result, only use throughput (inferences per second) for stabilizing. PA quite quickly with these updates.

Before changes:
DenseNet
![image](https://github.com/triton-inference-server/client/assets/58150256/a9045089-12d6-4ad5-8070-a39249c5a31b)


After changes:
DenseNet
![image](https://github.com/triton-inference-server/client/assets/58150256/9e350e84-ab0b-4e57-8bae-e772de0c5426)

tgerdesnv reviewed (commented):
code looks correct. would be nice to have some sort of testing to confirm and protect.

tgerdesnv commented:
There are also potential documentation changes needed?

tgerdesnv commented:
I'm also concerned about the fact that we still are reporting latency numbers. We may need a way to indicate that the latency is not stable

dyastremsky commented:
> code looks correct. would be nice to have some sort of testing to confirm and protect.

Correct in what way? You can see the screenshots above. I'll also be adding testing to ensure PA stabilizes to the main feature branch as a separate PR.

dyastremsky commented:
> I'm also concerned about the fact that we still are reporting latency numbers. We may need a way to indicate that the latency is not stable

Do you think this is something users care about? Will defer to you and @nnshah1 here. It seemed like even our team wasn't super familiar with what stabilization means, so I'm not sure that it's important enough to call out that latency may not have stabilized.

As a counterargument to the point about users not caring, we know that uncapping the request rate in this way can lead to unbounded growth so average latency won't be the most accurate measure. But we do show the latency per run, so users can see that.

nnshah1 commented:
> > I'm also concerned about the fact that we still are reporting latency numbers. We may need a way to indicate that the latency is not stable
> 
> Do you think this is something users care about? Will defer to you and @nnshah1 here. It seemed like even our team wasn't super familiar with what stabilization means, so I'm not sure that it's important enough to call out that latency may not have stabilized.
> 
> As a counterargument to the point about users not caring, we know that uncapping the request rate in this way can lead to unbounded growth so average latency won't be the most accurate measure. But we do show the latency per run, so users can see that.

My preference would be to handle this in documentation and not tool output.

If I understand correctly - we are saying that for cases were request rate >> throughput - latency is effectively unbounded - I think that's reasonable to document. Probably we want to add (if not present) the ability to sweep - i.e. start at low request rate and increase until throughput no longer increases - then stabilize latency at that request rate ....

tgerdesnv commented:
> > code looks correct. would be nice to have some sort of testing to confirm and protect.
> 
> Correct in what way? You can see the screenshots above. I'll also be adding testing to ensure PA stabilizes to the main feature branch as a separate PR.

I'm just commenting that the code changes look fine/correct to me.

dyastremsky commented:
> > > I'm also concerned about the fact that we still are reporting latency numbers. We may need a way to indicate that the latency is not stable
> > 
> > 
> > Do you think this is something users care about? Will defer to you and @nnshah1 here. It seemed like even our team wasn't super familiar with what stabilization means, so I'm not sure that it's important enough to call out that latency may not have stabilized.
> > As a counterargument to the point about users not caring, we know that uncapping the request rate in this way can lead to unbounded growth so average latency won't be the most accurate measure. But we do show the latency per run, so users can see that.
> 
> My preference would be to handle this in documentation and not tool output.
> 
> If I understand correctly - we are saying that for cases were request rate >> throughput - latency is effectively unbounded - I think that's reasonable to document. Probably we want to add (if not present) the ability to sweep - i.e. start at low request rate and increase until throughput no longer increases - then stabilize latency at that request rate ....

Created a ticket to look into sweeping the request rate ([TPA-140](https://jirasw.nvidia.com/browse/TPA-140)).

I added a warning message when the latency is not stabilizing, as discussed (see PR description). I also added a unit test. I can look at adding more extensive testing and the 0 sec avg HTTP time bug as separate PRs into the feature branch.

This pull request is ready for review.

nnshah1 reviewed (approved):
LGTM - defer to @tgerdesnv , @matthewkotila

debermudez commented on src/c++/perf_analyzer/inference_profiler.cc:
```diff
@@ -816,7 +827,8 @@ InferenceProfiler::ProfileHelper(
 bool
-InferenceProfiler::DetermineStability(LoadStatus& load_status)
+InferenceProfiler::DetermineStability(
+    LoadStatus& load_status, bool check_latency)
```
When is check_latency ever false?

dyastremsky replied:
We decided not to require latency to stabilize for the async case given the variability of latency in that case. If latency does not stabilize (but throughput does) by the end of the runs, then we want to print an error but still allow PA to stabilize.

See this line for more information: https://github.com/triton-inference-server/client/pull/688/files#diff-1d869f26de26a6908d10260f21bd433d63ab260cd58156aed650f563001547d1R796

debermudez replied:
i missed DetermineStability passing in false on that line. Thank you.
Would lines 796 - 802 technically be considered determining stability? Should that be internalized to the method and remove check_latency as an arg, atleast to the DetermineStability method?

dyastremsky replied:
It's still determining stability, just only of throughput. Are you recommending creating another DetermineStability function that only checks throughput?

dyastremsky replied:
Note that we still call DetermineStability for async models. If it doesn't ever stabilize, then we check that throughput stabilizes... if so, we allow it while printing a warning to the user.

debermudez replied:
I was thinking this logic around that space felt like it was all part of DeterminingStability.
I think we have 2 levels of checks now.
-first order: determine stability (called each pass?)
-second order: if not stable but in async mode, try again ignoring latency.

if i understand it correctly, what you are doing seems correct. Just feels like its all part of determiningStability.

This is probably highlighting a need for a refactor at this point, which we knew for PA already.
I think we can close this comment as you clearly thought this through. Just feels like every addition makes this code base harder to follow.
